lib/seahorn/Houdini.cc

```diff
@@ -0,0 +1,292 @@
+	unsigned i = 0;
+	for (i=0; i < bind::domainSz(fdecl); i++)
+	{
+		// if its type is INT
+		if (isOpX<INT_TY>(bind::domainTy(fdecl, i)))
+		{
+			// what is efac?
+			Expr bvar = bind::bvar (i, mk<INT_TY>(bind::domainTy(fdecl, i)->efac())); //the id of bvar is the same as related arg index
+			bvars.push_back(bvar);
+			bvar_count ++;
+		}
+	}
+
+	//What if there's no bvar?
+	if(bvar_count == 0)
+	{
+		cand = mk<TRUE>(fdecl->efac());
+	}
+	// if there is only one bvar, create a int constant and make an lt expr
+	else if(bvar_count == 1)
+	{
+		Expr zero = mkTerm<mpz_class> (0, fdecl->efac());
+		cand = mk<LT>(bvars[0], zero);
+	}
+	// if there are more than two bvars, make an lt expr
+	else if(bvar_count == 2)
+	{
+		Expr lt1 = mk<LT>(bvars[0], bvars[1]);
+		Expr lt2 = mk<LT>(bvars[1], bvars[0]);
+		bins.push_back(lt1);
+		bins.push_back(lt2);
+
+		cand = mknary<AND>(bins.begin(), bins.end());
+	}
+	else // bvar_count > 2
+	{
+		for(int j=0; j<bvars.size()-1; j++)
+		{
+			Expr lt = mk<LT>(bvars[j], bvars[j+1]);
+			bins.push_back(lt);
+		}
+		cand = mknary<AND>(bins.begin(), bins.end());
+	}
+
+	return cand;
+  }
+
+  #define SAT true
+  #define UNSAT false
+
+  std::map<Expr, Expr> Houdini::currentCandidates;	//a map from predicates to candidates
+
+  /*
+   * Build the map from predicates to candidates
+   */
+  void Houdini::guessCandidate(HornClauseDB &db)
+  {
+	  for(Expr rel : db.getRelations())
+	  {
+		  if(bind::isFdecl(rel))
+		  {
+			  Expr cand = relToCand(rel);
+			  currentCandidates.insert(std::pair<Expr, Expr>(rel, cand));
+		  }
+	  }
+  }
+
+  /*
+   * Main loop of Houdini algorithm
+   */
+  void Houdini::runHoudini(HornClauseDB &db, HornifyModule &hm)
+  {
+	  guessCandidate(db);
+
+	  HornClauseDB::RuleVector workList;
+	  workList.insert(workList.end(), db.getRules().begin(), db.getRules().end());
+
+	  while(!workList.empty())
+	  {
+		  HornRule r = workList.front();
+		  errs() << "RULE HEAD: " << *(r.head()) << "\n";
+		  errs() << "RULE BODY: " << *(r.body()) << "\n";
+		  while (validateRule(r, db, hm) == SAT)
+		  {
+			  workList = addUsedRulesBackToWorkList(db, workList, r.head());
+			  weakenRuleHeadCand(r);
+		  }
+		  workList.erase(workList.begin());
+	  }
+  }
+
+  /*
+   * Given a rule, check if it's validate
+   */
+  bool Houdini::validateRule(HornRule r, HornClauseDB &db, HornifyModule &hm)
+  {
+	  Expr ruleHead_cand_app = applyActualArgsToCand(r.head());
+	  Expr neg_ruleHead_cand_app = mk<NEG>(ruleHead_cand_app);
+
+	  Expr ruleBody = r.body();
+	  ExprMap body_map;
+	  ExprVector body_pred_apps;
+	  get_all_pred_apps(ruleBody, db, std::back_inserter(body_pred_apps));
+
+	  for(ExprVector::iterator itr = body_pred_apps.begin(); itr != body_pred_apps.end(); ++itr)
+	  {
+		  body_map.insert(std::pair<Expr, Expr>(*itr, applyActualArgsToCand(*itr)));
+	  }
+
+	  Expr body_cand_app = replace(ruleBody, body_map);
+
+	  Expr whole_cand = mk<AND>(neg_ruleHead_cand_app, body_cand_app);
```

Comment:
Thanks Arie. I think I misunderstand the design of weakenCandidate(), So I hope to ask if solver.assertExpr(e) says e is SAT. Then should I use getModel() to get the model that makes the results true? Then how do I get the lemma that I should remove from the model? Thanks.